dev/tests/integration/testsuite/Magento/Downloadable/Model/File/ContentUploaderTest.php

```diff
@@ -0,0 +1,118 @@
+use Magento\TestFramework\Helper\Bootstrap;
+use PHPUnit\Framework\MockObject\MockObject;
+use PHPUnit\Framework\TestCase;
+
+/**
+ * Integration test for \Magento\Downloadable\Model\File\ContentUploader class
+ */
+class ContentUploaderTest extends TestCase
+{
+    /**
+     * @var MockObject
+     */
+    private $fileContentMock;
+
+    /**
+     * @var string
+     */
+    private $filePath;
+
+    /**
+     * @var ContentUploader
+     */
+    private $model;
+
+    /**
+     * @var ObjectManagerInterface
+     */
+    private $objectManager;
+
+    /**
+     * @inheritdoc
+     */
+    protected function setUp(): void
+    {
+        $this->objectManager = Bootstrap::getObjectManager();
+        $this->fileContentMock = $this->objectManager->create(ContentInterfaceFactory::class)->create();
+        $fixtureDir = realpath(__DIR__ . '/../../_files');
+        $this->filePath = $fixtureDir . DIRECTORY_SEPARATOR . 'test_image.jpg';
+        $this->fileContentMock->setFileData(base64_encode(file_get_contents($this->filePath)));
+        $this->fileContentMock->setName('test_image.jpg');
+
+        $database = $this->getMockBuilder(Database::class)
+            ->disableOriginalConstructor()
+            ->getMock();
+        $storage = $this->getMockBuilder(Storage::class)
+            ->disableOriginalConstructor()
+            ->getMock();
+        $validator = $this->getMockBuilder(NotProtectedExtension::class)
+            ->disableOriginalConstructor()
+            ->getMock();
+        $filesystem = $this->getMockBuilder(Filesystem::class)
+            ->disableOriginalConstructor()
+            ->onlyMethods(['getDirectoryWrite'])
+            ->getMock();
+        $systemTmpDirectory = $this->getMockForAbstractClass(
+            WriteInterface::class,
+            [],
+            '',
+            false,
+            true,
+            true,
+            ['writeFile']
+        );
+        $systemTmpDirectory->expects($this->once())->method('writeFile')->willReturn(1);
+        $systemTmpDirectory->method('getAbsolutePath')->willReturn($this->filePath);
+
+        $filesystem->method('getDirectoryWrite')->willReturn($systemTmpDirectory);
+        $link = $this->getMockBuilder(Link::class)
+            ->disableOriginalConstructor()
+            ->getMock();
+        $sampleConfig = $this->getMockBuilder(Sample::class)
+            ->disableOriginalConstructor()
+            ->getMock();
+        $this->model = $this->getMockForAbstractClass(
+            ContentUploader::class,
+            [$database, $storage, $validator, $filesystem, $link, $sampleConfig],
+            '',
+            true,
+            true,
+            true,
+            ['save']
+        );
+    }
+
+    public function testUploadWithSuccessSave()
+    {
+        $data = ['path' => $this->filePath, 'file' => 'test_image.jpg'];
+        $this->model->expects($this->once())->method('save')->willReturn($data);
```

Comment:
Shouldn't this test be converted to unit test with all those mocks ?

Reply:
Hi @xmav , I think it's easier to leave it as an integration test.
It performs a check on a real image, and it's convenient to use Bootstrap from `dev/tests/integration/framework/Magento/TestFramework/Helper/Bootstrap.php`
to create a factory
https://github.com/magento/magento2/pull/34616/files#diff-35e356eb4f93c785d13631985cd91beaa65f19a01012aa40dad7cc2552129126R56